### PR TITLE
[Pooling] Fix normalization of chunked long-text embeddings

### PR DESCRIPTION
**Pull request by taneem-ibrahim**
## Purpose

Averaging normalized chunks can produce a non-unit long-text embedding. Re-normalize multi-chunk aggregates when `use_activation=True`, preserving single-chunk and raw outputs.

## Reproducer

The temporary CPU reproducer passes two equal-weight, orthogonal normalized
chunk embeddings through `_post_process_chunked`. It also checks a one-chunk
BF16 prompt and the `use_activation=False` path.

```bash
env PYTHONPATH=<main-worktree> .venv/bin/python /tmp/chunked_norm_cpu_reproducer.py
env PYTHONPATH=$PWD .venv/bin/python /tmp/chunked_norm_cpu_reproducer.py
```

## Output on main / on branch

Main:

```text
multi=[0.5, 0.5] norm=0.707107
single_exact=True value=[0.6015625, 0.80078125]
raw=[0.5, 0.5]
```

Branch:

```text
multi=[0.7071067690849304, 0.7071067690849304] norm=1.000000
single_exact=True value=[0.6015625, 0.80078125]
raw=[0.5, 0.5]
```

## H100 multi-model evaluation

The  end-to-end sweep used BF16, `max_model_len=128`, and long inputs of 1,192-1,682 tokens. It covered encoder mean pooling, encoder CLS pooling, and decoder last-token pooling.

| Model | Architecture / pooling | Old multi-chunk norms | New multi-chunk norms | Short max abs delta across independent servers | Minimum cosine with old output | Retrieval ranking |
| --- | --- | ---: | ---: | ---: | ---: | --- |
| `intfloat/e5-small-v2` | encoder / MEAN | 0.997116-0.997419 | 1.000000 | 7.45e-9 | 1.000000 | cosine and IP unchanged |
| `BAAI/bge-small-en-v1.5` | encoder / CLS | 0.986880-0.993246 | 1.000000 | 1.49e-8 | 1.000000 | cosine and IP unchanged |
| `Qwen/Qwen3-Embedding-0.6B` | decoder / LAST | 0.784638-0.821176 | 1.000000 | 7.45e-9 | 0.999994 | cosine unchanged; IP changed |

For E5 and BGE, the new long output matched an explicitly normalized old weighted mean within `2.98e-8`. Qwen's maximum BF16 difference was `4.73e-4` with the same cosine ranking. Its IP ranking changed from `[0, 1, 2, 4, 3]`
to `[0, 1, 4, 2, 3]`, demonstrating why this should be documented as an intentional correctness-related output change rather than a refactor.

Short prompts do not execute the new normalization operation. The direct BF16 aggregation check is bitwise exact; independent H100 server launches differed by at most `1.49e-8`, consistent with normal numerical variation.

### Compatibility note

Only normalized outputs formed from multiple chunks change. Their direction is preserved while their magnitude becomes approximately 1. Cosine retrieval is therefore expected to remain stable, but inner-product scores and rankings can change. Deployments using inner-product indexes should not mix old and new long-document embeddings; re-embed affected long documents when upgrading.

## Test plan

```bash
.venv/bin/python -m pytest tests/entrypoints/pooling/embed/test_io_processor.py -q
# 46 passed

env UV_CACHE_DIR=/tmp/vllm-uv-cache .venv/bin/pre-commit run \
  --files vllm/entrypoints/pooling/embed/io_processor.py
# Passed

env UV_CACHE_DIR=/tmp/vllm-uv-cache .venv/bin/pre-commit run \
  --all-files --hook-stage manual
# Passed, including dependency locks and mypy for Python 3.10-3.13
```

The deterministic temporary CPU reproducer and the three-model H100 E2E sweep also passed with the outputs summarized above.

## AI assistance disclosure

OpenAI Codex (GPT-5) assisted with drafting the implementation and temporary validation scripts.

### CODE REVIEW

File: vllm/entrypoints/pooling/embed/io_processor.py [hunk 1]
```diff
@@ -239,6 +241,10 @@ def _post_process_chunked(self, ctx: PoolingServeContext) -> None:
                 ):
                     # Compute final mean embedding
                     final_embedding = weighted_sum / total_weight
+                    if ctx.pooling_params.use_activation and aggregator.num_chunks > 1:
+                        final_embedding = torch.nn.functional.normalize(
+                            final_embedding, dim=-1
+                        )
```
Comment (yewentao256) [inline vllm/entrypoints/pooling/embed/io_processor.py]: Please check if `use_activation` guarantee L2-normalized?

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!

File: vllm/entrypoints/pooling/embed/io_processor.py (hunk 1 above)
Comment (taneem-ibrahim) [reply vllm/entrypoints/pooling/embed/io_processor.py]: Good catch! `use_activation` does not guarantee L2 normalization. I updated the aggregation to normalize only when every actual chunk output is unit norm.

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!
